xdsl/frontend/passes/desymref.py

```diff
@@ -0,0 +1,389 @@
+        """
+        Promotes an operation. This method guarantees that the operation does
+        not have any symbols.
+        """
+        # TODO: Add promoters in the next patch.
+        pass
+
+    def prepare_op(self, op: Operation):
+        """
+        Prepares an operation for promotion. This method guarantees that any
+        symbol in any region of this operation is read at most once and written
+        at most once.
+        """
+
+        # For operation with no regions we don't have to do any work.
+        if len(op.regions) == 0:
+            return
+
+        # Otherwise, we have to prepare regions.
+        for region in op.regions:
+            self.prepare_region(region)
+
+    def prepare_region(self, region: Region):
+        """
+        Prepares a region for promotion. This method guarantees that any symbol
+        in the region is read at most once and written at most once.
+        """
+        num_blocks = len(region.blocks)
+        if num_blocks == 1:
+            # If there is only one block, preparing region is easier, so we
+            # handle it separately.
+            self.prepare_block(region.blocks[0])
+        else:
+            # TODO: Support regions with multiple blocks. This is not too
+            # difficult but requires many more things:
+            # 1. SSA-construction algorithm to prune symbol declarations. This
+            #    also requires analyses passes (dominators).
+            # 2. Insertion of entry/exit blocks to ensure dominance.
+            raise FrontendProgramException(
+                f"Running desymrefier on region with {num_blocks} > 1 blocks is "
+                "not supported.")
+
+    def prepare_block(self, block: Block):
+        """Prepares a block for promotion."""
+
+        # First, desymrefy nested regions.
+        for op in block.ops:
+            self.desymrefy(op)
+
+        self.prune_definitions(block)
+        self.prune_uses_without_definitions(block)
+
+        # Ensure that each symbol is read/written at most once.
+        symbols = get_symbols(block)
+        for symbol in symbols:
+            num_reads = count_ops_by(
+                block, lambda op: is_read(op) and get_symbol(op) == symbol)
+            num_writes = count_ops_by(
+                block, lambda op: is_write(op) and get_symbol(op) == symbol)
+            if num_reads > 1 or num_writes > 1:
+                raise FrontendProgramException(
+                    f"Block {block} not ready for promotion: found {num_reads}"
+                    f" reads and {num_writes} writes.")
+
+    def prune_definitions(self, block: Block):
+        """Removes all symbol definitions and their uses from the block."""
+        while True:
+            # Find all symbol definitions in this block. If no definitions
+            # found, terminate.
+            definitions = select_ops_by(block, is_definition)
+            if len(definitions) == 0:
+                return
```

Comment:
I still don't quite like the coding style here. IMO, `while True` is not great style, I would prefer to have a do-while loop or duplicate these two lines before the loop.

Reply:
Maybe you can outline the condition checking into a function, that returns None instead of 0? Than you could easily use a walrus for example...

Reply:
Not sure that is better coding style at this point xD

Reply:
I think you write do-while loops this way in Python though... But same, I do not know what coding style we should use here 😅

Reply:
`while len(definitions := [op for op in block.ops if isinstance(op, symref.Declare)]) > 0:` ?
That would be ~~pythonic~~